### PR TITLE
feat: FDC3 integration GENC-224

### PR DESCRIPTION
**Pull request by patrickoneill-genesis**
[Review Guidance](https://www.notion.so/genesisglobal/Platform-Code-Review-Process-ace93ba760cc4563b0dfb712e2a88d8a)

📓 &nbsp; **Related JIRA**

[PTC-0](https://genesisglobal.atlassian.net/browse/PTC-0)

🤔 &nbsp; **What does this PR do?**
Add fdc3 channel listening config to the genx blank app seed 

🚀 &nbsp; **Where should the reviewer start?**

Add file / directory pointers.

📑 &nbsp; **How should this be tested?**

If you have an fdc3 agent, create this app, and then another app that sends a message on the channel you specify
Run the following command and it will generate a main.ts with a channel listener

`TESTAPP=prtestapp && \
                BLANK_APP_SEED_PATH=./blank-app-seed && \
                rm -rf $TESTAPP && \
                npx -y @genesislcap/genx@latest init $TESTAPP -x --fdc3 -s $BLANK_APP_SEED_PATH --apiHost 'wss://public-foundation.genesislab.global/gwf/' --no-npm --no-shell --fdc3ListenChannel "{\"name\":\"testChanne\",\"type\":\"testType\"}" --routes "[{\"name\":\"home\",\"tiles\":[{\"title\":\"Grid Tile\",\"type\":\"grid-pro\",\"config\":{\"resourceName\":\"ALL_POSITIONS\"}},{\"title\":\"Form tile\",\"type\":\"smart-form\",\"config\":{\"resourceName\":\"EVENT_COUNTERPARTY_INSERT\"}}]}]" && \
                cd $TESTAPP/client && \
                npm run bootstrap && npm run dev`

```
npx -y @genesislcap/genx@latest init prtest -x --ref YOUR-BRANCH-NAME --no-npm
```

✅ &nbsp; **Checklist**

- [x] I have tested my changes.
- [ ] I have added tests for my changes.
- [ ] I have updated the project documentation to reflect my changes.

### CODE REVIEW

**Comment by patrickoneill-genesis** [comment]
```
TESTAPP=prtestapp && \
                BLANK_APP_SEED_PATH=./blank-app-seed && \
                rm -rf $TESTAPP && \
                npx -y @genesislcap/genx@latest init $TESTAPP -x --fdc3 -s $BLANK_APP_SEED_PATH --apiHost 'wss://public-foundation.genesislab.global/gwf/' --no-npm --no-shell --fdc3ListenChannel "{\"name\":\"testChanne\",\"type\":\"testType\"}" --routes "[{\"name\":\"home\",\"tiles\":[{\"title\":\"Grid Tile\",\"type\":\"grid-pro\",\"config\":{\"resourceName\":\"ALL_POSITIONS\", \"fdc3RowClick\":{\"channel\":\"rowClickChannel\",\"type\":\"rowClickType\"}}},{\"title\":\"Grid Tile 2\",\"type\":\"entity-manager\",\"config\":{\"resourceName\":\"ALL_POSITIONS\", \"fdc3RowClick\":{\"channel\":\"rowClickChannel2\",\"type\":\"rowClickType2\"}}},{\"title\":\"Form tile\",\"type\":\"smart-form\",\"config\":{\"resourceName\":\"EVENT_COUNTERPARTY_INSERT\"}}]}]" && \
                cd $TESTAPP/client && \
                npm run bootstrap && npm run dev
```
                
                Create a grid and entity manager with row click listeners.

File: .idea/workspace.xml [hunk 1]
```diff
@@ -3,6 +3,66 @@
   <component name="AutoImportSettings">
     <option name="autoReloadType" value="SELECTIVE" />
   </component>
```
Comment (cistov) [inline .idea/workspace.xml:5]: Did you intend to commit this?

File: .genx/templates/main/main.hbs [hunk 2]
```diff
@@ -48,8 +52,12 @@ export class MainApplication extends EventEmitter<StoreEventDetailMap>(FASTEleme
...
     });
 
-    {{#if fdc3}}
-    console.log('FDC3 listening logic here');
+    {{#if data.fdc3ListenChannel}}
+        if (window.fdc3) {
+            await this.fdcReadyHandler();
+        } else {
+            window.addEventListener('fdc3Ready', async () => this.fdcReadyHandler());
+        }
```
Comment (cistov) [inline .genx/templates/main/main.hbs:60]: We could extract this into another little helper in fdc3 utils:

```
{{#if data.fdc3ListenChannel}}
onFDC3Ready(this.fdcReadyHandler);
{{/if}}
```

File: .genx/templates/main/main.hbs [hunk 3]
```diff
@@ -1,3 +1,6 @@
+{{#if data.fdc3ListenChannel}}
+import { listenToChannel } from '../utils';
+{{/if}}
 import { Connect, ConnectConfig, defaultConnectConfig } from '@genesislcap/foundation-comms';
```
Comment (cistov) [inline .genx/templates/main/main.hbs:4]: This rename/move shouldn't be needed - I can sort out in my branch though

Comment (patrickoneill-genesis) [reply .genx/templates/main/main.hbs:4]: It's not just `fdc3` it's a more specific action

File: .genx/templates/main/main.hbs (hunk 2 above)
Comment (patrickoneill-genesis) [reply .genx/templates/main/main.hbs:60]: listen to channel is already abstracted - you might need access to `this` or something in the class.

File: .genx/templates/grid.hbs [hunk 4]
```diff
@@ -1,9 +1,20 @@
...
     {{/if}}
+      {{#if config.fdc3RowClick}}
+    :deferredGridOptions="${() => {
+            return {
+                onRowClicked: async (e: RowSelectedEvent) => {
+                    const payload = e.data
+                    sendMessageOnChannel('{{config.fdc3RowClick.channel}}', '{{config.fdc3RowClick.type}}', payload)
+                }
+            }
+        }}"
```
Comment (cistov) [inline .genx/templates/grid.hbs:16]: If we add a util similar to this (basically a getter for the event handler function):

```ts
export const sendEventOnChannel = (channelName: string, type: string) => {
    return async (e: any) => {
        // check for ag-grid-specific events, fall back to standard events
        const payload = e.data?.payload || e.detail;
        sendMessageOnChannel(channelName, type, payload);
    }
}
```

we could simplify the configuration block to just:

```ts
:deferredGridOptions="${() => ({
  onRowClicked: sendEventOnChannel('{{config.fdc3RowClick.channel}}', '{{config.fdc3RowClick.type}}')
})}"
```

File: .genx/prompts/ui.js [hunk 5]
```diff
@@ -24,8 +25,27 @@ module.exports = async (inquirer, prevAns = {}) => {
       when: !prevAns.fdc3,
       default: prevAns.fdc3 || false,
     },
+    {
+      name: 'fdc3ListenChannel',
+      type: 'confirm',
+      message: 'Listen to an fdc channel config in json format',
+      when: !prevAns.fdc3ListenChannel,
+      default: prevAns.fdc3ListenChannel || false,
+    },
```
Comment (cistov) [inline .genx/prompts/ui.js:34]: we'll need to rejig these prompts a bit, but I can do this once it's merged into my branch

File: .genx/templates/main/main.hbs (hunk 3 above)
Comment (cistov) [reply .genx/templates/main/main.hbs:4]: I meant moving of `client/src/main/main.ts` to `.genx/templates/main/main.hbs`.

File: .genx/templates/main/main.hbs (hunk 2 above)
Comment (cistov) [reply .genx/templates/main/main.hbs:60]: I meant subscribing to fdc3 ready event - there's some boilerplate we can do away with. You'll still have access to this if needed.

File: .genx/templates/grid.hbs (hunk 4 above)
Comment (patrickoneill-genesis) [reply .genx/templates/grid.hbs:16]: that's very grid specific so will add another method

File: .genx/templates/main/main.hbs (hunk 2 above)
Comment (patrickoneill-genesis) [reply .genx/templates/main/main.hbs:60]: Yes, but need to pass the function like `() => this.fdcReadyHandler()`

File: .idea/workspace.xml (hunk 1 above)
Comment (patrickoneill-genesis) [reply .idea/workspace.xml:5]: No, will revert this

**Review by cistov — APPROVED** [review]

**Pull request merged**